### ISSUE TITLE
Need better testing story for anonymous workers

### ISSUE THREAD

**Issue by zach-klippenstein** [square/workflow#489]
It is very convenient to create anonymous workers inside a workflow. However, since they're not injected, they can't be tested with `testRender`. They _can_ be tested with `testFromState`, but only starting state can be specified – there's no way to get the transitioned-to state. 

Should we make `WorkflowTester` expose state transitions?

cc @rjrjr @grantamos

**Comment by zach-klippenstein** [comment]
We can provide an API to create fake placeholder workers that the test library uses to compare specifically only on keys, instead of using `doesSameWorkAs` like normal workers.

Test library API:
```kotlin
fun matchingWorker(key: String): Worker<Nothing>
```

Example usage:
```kotlin
class MyWorkflow : StatelessWorkflow<Unit, Nothing, Unit>() {
  private val anonymousWorker = Worker.create<String> { … }

  fun render(…) {
    …
    context.onWorkerOutput(anonymousWorker, "anonKey") { … }
    …
  }
}

class MyTest {
  private val workflow = MyWorkflow()
  private val worker = matchingWorker<String>()

  @Test fun test() {
    workflow.testRenderInitialState {
      worker.assertRan(withKey = "anonKey")
    }
  }
}
```

**Comment by zach-klippenstein** [comment]
In order to support matching anonymous workers, we need a way to _identify_ anonymous workers. This isn't an issue for custom workers – since they have a specific concrete type, a matcher predicate can just do an `is` check on that type. Anonymous workers (or at least all the anonymous workers we currently implement, which are all used to implement conversions from other reactive/async types) are now only distinguished by their output type. All anonymous workers provided by the library are currently `TypedWorker` under the hood, which contains the `KClass` of the output type in order to implement `doesSameWorkAs`. In order to write a matcher for one of these workers, you'd need to be able to compare the reified output types.

There are a few ways we can do this:
1. Make `runningWorker` take a `KClass`<sup>1</sup> with an inline extension for reification and associate the type with the worker itself. We actually used to do this when we first scrapped Reactor and moved to the new style of workflows. The issues with this are exposing extra API (`RenderContext` needs a version of `runningWorker` that nobody should ever call directly), and that it prevents workflows from being generic on a worker type – a generic class can't use its own type parameters for passing to inline reified functions.
2. Introduce a public `TypedWorker` sub-interface of `Worker` that contains a `val outputType: KClass<OutputT>`. Custom workers wouldn't need to, but all anonymous workers should then implement this interface, and we could provide a matching helper for tests like `fun isWorkerOfType<T>() = worker is TypedWorker && worker.outputType == T::class`. The main issue with this approach is that tests now have to be aware of the existence of `TypedWorker`, and know whether or not a particular worker is actually implemented as a typed worker under the hood. As someone writing a workflow and then testing it, this detail is going to feel arbitrary, and worse, could potentially change (workflow goes from using an anonymous worker to a custom one, and now the test has to change how it mocks the worker for no good reason). The existence of `TypedWorker` is also completely irrelevant when actually writing a workflow, so it's extra weird to make people only think about it when they're writing tests.
3. Hang the `val outputType` off the main `Worker` interface directly. This means that the core `Worker` contract now includes the reification of its output type, but that's good because it means tests can be written consistently to identify workers by their output type, regardless of their implementation. It could be a little redundant for custom workers, but also means tests matching on custom workers could also be consistent with matching on anonymous workers. It's also a little more boilerplate for custom workers, but not much: `override val outputType = String::class`. This option is also the closest to how Swift would likely implement anonymous workers if they had them – because Swift has reified generics, they effectively get this `outputType` property for free.

After some discussion with @rjrjr and @loganj we all agreed that option 3 is the best one, so I'll try to get around to building this for v0.21.0.

----
<sup>1</sup> Ideally would actually be `KType`, so we could distinguish between nullable/non-nullable types, higher-order generics, etc, but our build system (Buck) can't handle kotlin's reflection library so that's not an option at this time.
